common/routing/route.go

```diff
@@ -0,0 +1,170 @@
+}
+
+// RouteBuilder is a builder for the [Route] interface.
+type RouteBuilder[T any] struct {
+	components []Component[T]
+}
+
+// NewBuilder creates a new [RouteBuilder] instance, which can be used to define a new [Route] via a fluent API.
+func NewBuilder[T any]() *RouteBuilder[T] {
+	return &RouteBuilder[T]{}
+}
+
+// With adds a series of [Component] instances to the [Route].
+func (r *RouteBuilder[T]) With(c ...Component[T]) *RouteBuilder[T] {
+	r.components = append(r.components, c...)
+	return r
+}
+
+// Constant adds a [Constant] component to the [Route].
+func (r *RouteBuilder[T]) Constant(values ...string) *RouteBuilder[T] {
+	return r.With(Constant[T](values...))
+}
+
+// Variable adds a [Variable] component to the [Route].
+func (r *RouteBuilder[T]) Variable(name string, getter func(*T) *string) *RouteBuilder[T] {
+	return r.With(Variable[T](name, getter))
+}
+
+// Build returns a read-only [Route].
+func (r *RouteBuilder[T]) Build() Route[T] {
+	return NewRoute[T](r.components...)
+}
+
+// Representation returns the [github.com/gorilla/mux] compatible string representation of the route for usage in a
+// path definition. It does not add a leading or trailing slash to the representation, but it won't remove them if
+// they're present in the components. We do this because it's easier to add a slash depending on the context than to
+// remove it.
+func (r Route[T]) Representation() string {
+	return r.serialize(func(c Component[T]) string {
+		return c.Representation()
+	})
+}
+
+// Path returns the serialized path of the route with the given params. There will be no leading or trailing slashes,
+// similar to the behavior of the [Route.Representation] method.
+func (r Route[T]) Path(t T) string {
+	return r.serialize(func(c Component[T]) string {
+		return c.Serialize(t)
+	})
+}
+
+func (r Route[T]) serialize(f func(c Component[T]) string) string {
+	var sb strings.Builder
+	for i, c := range r.components {
+		if i > 0 {
+			sb.WriteString("/")
+		}
+		sb.WriteString(f(c))
+	}
+	return sb.String()
+}
+
+// Deserialize the given vars into a new instance of the params type, T.
+func (r Route[T]) Deserialize(vars map[string]string) *T {
+	var t T
+	for _, c := range r.components {
+		c.Deserialize(vars, &t)
+	}
+	return &t
+}
+
+// Constant returns a [Component] that represents a series of constant HTTP path components in a Route.
+// They will be joined via strings when used to construct a path or path representation.
+func Constant[T any](values ...string) constants[T] {
+	return values
+}
+
+type constants[T any] []string
+
+func (s constants[T]) Representation() string {
+	return strings.Join(s, "/")
+}
+
+func (s constants[T]) Serialize(T) string {
+	return strings.Join(s, "/")
+}
+
+func (s constants[T]) Deserialize(map[string]string, *T) {}
+
+// Variable returns a [Component] that represents a string variable in a Route.
+func Variable[T any](name string, getter func(*T) *string) stringVariable[T] {
```

Comment:
nit: This should have been called `StringVariable` IMHO so we can support other variable types.